File: play-ahc-ws-standalone/src/main/scala/play/api/libs/ws/ahc/StandaloneAhcWSClient.scala
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws.ahc

import akka.Done
import javax.inject.Inject
import akka.stream.Materializer
import akka.stream.scaladsl.Source
import akka.util.ByteString
import com.typesafe.sslconfig.ssl.SystemConfiguration
import org.reactivestreams.Publisher
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription
import play.api.libs.ws.ahc.cache._
import play.api.libs.ws.EmptyBody
import play.api.libs.ws.StandaloneWSClient
import play.api.libs.ws.StandaloneWSRequest
import play.shaded.ahc.org.asynchttpclient.uri.Uri
import play.shaded.ahc.org.asynchttpclient.{ Response => AHCResponse }
import play.shaded.ahc.org.asynchttpclient._
import java.util.function.{ Function => JFunction }

import scala.collection.immutable.TreeMap
import scala.jdk.FunctionConverters._
import scala.concurrent.Await
import scala.concurrent.Future
import scala.concurrent.Promise
import scala.util.Failure
import scala.util.Success

/**
 * A WS client backed by an AsyncHttpClient.
 *
 * If you need to debug AsyncHttpClient, add <logger name="play.shaded.ahc.org.asynchttpclient" level="DEBUG" /> into your conf/logback.xml file.
 *
 * @param asyncHttpClient An already configured asynchttpclient.
 *                        Note that the WSClient assumes ownership of the lifecycle here, so closing the WSClient will
 *                        also close asyncHttpClient.
 * @param materializer A materializer, meant to execute the stream
 */
class StandaloneAhcWSClient @Inject() (asyncHttpClient: AsyncHttpClient)(implicit
    materializer: Materializer
) extends StandaloneWSClient {

  /** Returns instance of AsyncHttpClient */
  def underlying[T]: T = asyncHttpClient.asInstanceOf[T]

  def close(): Unit = {
    asyncHttpClient.close()
  }

  def url(url: String): StandaloneWSRequest = {
    validate(url)
    StandaloneAhcWSRequest(
      client = this,
      url = url,
      method = "GET",
      body = EmptyBody,
      headers = TreeMap()(CaseInsensitiveOrdered),
      queryString = Map.empty,
      cookies = Seq.empty,
      calc = None,
      auth = None,
      followRedirects = None,
      requestTimeout = None,
      virtualHost = None,
      proxyServer = None,
      disableUrlEncoding = None
    )
  }

  private[ahc] def execute(
      request: Request
  ): Future[StandaloneAhcWSResponse] = {
    val result = Promise[StandaloneAhcWSResponse]()
    val handler = new AsyncCompletionHandler[AHCResponse]() {
      override def onCompleted(response: AHCResponse): AHCResponse = {
        result.success(StandaloneAhcWSResponse(response))
        response
      }

      override def onThrowable(t: Throwable): Unit = {
        result.failure(t)
      }
    }

    asyncHttpClient.executeRequest(request, handler)
    result.future
  }

  private def validate(url: String): Unit = {
    // Recover from https://github.com/AsyncHttpClient/async-http-client/issues/1149
    try {
      Uri.create(url)
    } catch {
      case iae: IllegalArgumentException =>
        throw new IllegalArgumentException(s"Invalid URL $url", iae)
      case npe: NullPointerException =>
        throw new IllegalArgumentException(s"Invalid URL $url", npe)
    }
  }

  private[ahc] def executeStream(request: Request): Future[StreamedResponse] = {
    val streamStarted    = Promise[StreamedResponse]()
    val streamCompletion = Promise[Done]()

    val client = this

    val function: JFunction[StreamedState, StreamedResponse] = { (state: StreamedState) =>
      val publisher = state.publisher

      val wrap = new Publisher[HttpResponseBodyPart]() {
        override def subscribe(
            s: Subscriber[_ >: HttpResponseBodyPart]
        ): Unit = {
          publisher.subscribe(new Subscriber[HttpResponseBodyPart] {
            override def onSubscribe(sub: Subscription): Unit =
              s.onSubscribe(sub)

            override def onNext(t: HttpResponseBodyPart): Unit = s.onNext(t)

            override def onError(t: Throwable): Unit = s.onError(t)

            override def onComplete(): Unit = {
              streamCompletion.future.onComplete {
                case Success(_) => s.onComplete()
                case Failure(t) => s.onError(t)
              }(materializer.executionContext)
            }
          })
        }

      }
      new StreamedResponse(
        client,
        state.statusCode,
        state.statusText,
        state.uriOption.get,
        state.responseHeaders,
        wrap,
        asyncHttpClient.getConfig.isUseLaxCookieEncoder
      )

    }.asJava
    asyncHttpClient.executeRequest(
      request,
      new DefaultStreamedAsyncHandler[StreamedResponse](
        function,
        streamStarted,
        streamCompletion
      )
    )
    streamStarted.future
  }

  private[ahc] def blockingToByteString(bodyAsSource: Source[ByteString, _]) = {
    StandaloneAhcWSClient.logger.warn(
      s"blockingToByteString is a blocking and unsafe operation!"
    )

    import scala.concurrent.ExecutionContext.Implicits.global

    val limitedSource = bodyAsSource.limit(StandaloneAhcWSClient.elementLimit)
    val result = limitedSource
      .runFold(ByteString.createBuilder) { (acc, bs) =>
        acc.append(bs)
      }
      .map(_.result())

    Await.result(result, StandaloneAhcWSClient.blockingTimeout)
  }

}

object StandaloneAhcWSClient {

  import scala.concurrent.duration._
  val blockingTimeout = 50.milliseconds
  val elementLimit    = 13 // 13 8192k blocks is roughly 100k
  private val logger  = org.slf4j.LoggerFactory.getLogger(this.getClass)

  private[ahc] val loggerFactory = new AhcLoggerFactory(
    org.slf4j.LoggerFactory.getILoggerFactory
  )

  /**
   * Convenient factory method that uses a play.api.libs.ws.WSClientConfig value for configuration instead of
   * an [[http://static.javadoc.io/org.asynchttpclient/async-http-client/2.0.0/org/asynchttpclient/AsyncHttpClientConfig.html org.asynchttpclient.AsyncHttpClientConfig]].
   *
   * Typical usage:
   *
   * {{{
   * import play.api.libs.ws.ahc.StandaloneAhcWSClient
   *
   * def example(someUrl: String)(implicit m: akka.stream.Materializer) = {
   *   implicit def ec = m.executionContext
   *
   *   val client = StandaloneAhcWSClient()
   *   val request = client.url(someUrl).get()
   *
   *   request.foreach { response =>
   *     //doSomething(response)
   *     client.close()
   *   }
   * }
   * }}}
   *
   * @param config configuration settings
   * @param httpCache if not null, will be used for HTTP response caching.
   * @param materializer the akka materializer.
   */
  def apply(
      config: AhcWSClientConfig = AhcWSClientConfigFactory.forConfig(),
      httpCache: Option[AhcHttpCache] = None
  )(implicit materializer: Materializer): StandaloneAhcWSClient = {
    val ahcConfig       = new AhcConfigBuilder(config).build()
    val asyncHttpClient = new DefaultAsyncHttpClient(ahcConfig)
    val wsClient = new StandaloneAhcWSClient(
      httpCache
        .map { cache =>
          new CachingAsyncHttpClient(asyncHttpClient, cache)
        }
        .getOrElse {
          asyncHttpClient
        }
    )
    new SystemConfiguration(loggerFactory).configure(config.wsClientConfig.ssl)
    wsClient
  }
}

File: play-ahc-ws-standalone/src/main/java/play/libs/ws/ahc/StandaloneAhcWSClient.java
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws.ahc;

import akka.Done;
import akka.stream.Materializer;
import akka.stream.javadsl.Source;
import akka.util.ByteString;
import akka.util.ByteStringBuilder;
import com.typesafe.sslconfig.ssl.SystemConfiguration;
import com.typesafe.sslconfig.ssl.debug.DebugConfiguration;
import org.reactivestreams.Publisher;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;
import org.slf4j.LoggerFactory;
import play.api.libs.ws.ahc.*;
import play.api.libs.ws.ahc.cache.AhcHttpCache;
import play.api.libs.ws.ahc.cache.CachingAsyncHttpClient;
import play.libs.ws.StandaloneWSClient;
import play.libs.ws.StandaloneWSResponse;
import play.shaded.ahc.org.asynchttpclient.AsyncCompletionHandler;
import play.shaded.ahc.org.asynchttpclient.AsyncHttpClient;
import play.shaded.ahc.org.asynchttpclient.DefaultAsyncHttpClient;
import play.shaded.ahc.org.asynchttpclient.DefaultAsyncHttpClientConfig;
import play.shaded.ahc.org.asynchttpclient.Request;
import play.shaded.ahc.org.asynchttpclient.Response;
import play.shaded.ahc.org.asynchttpclient.*;
import scala.compat.java8.FutureConverters;
import scala.compat.java8.FutureConverters$;
import scala.concurrent.ExecutionContext;
import scala.concurrent.Future;
import scala.concurrent.Promise;

import javax.inject.Inject;
import java.io.IOException;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import java.util.function.Function;

/**
 * A WS asyncHttpClient backed by an AsyncHttpClient instance.
 */
public class StandaloneAhcWSClient implements StandaloneWSClient {

    private final AsyncHttpClient asyncHttpClient;
    private final Materializer materializer;

    /**
     * Creates a new client.
     *
     * @param asyncHttpClient the underlying AsyncHttpClient
     * @param materializer    the Materializer to use for streams
     */
    @Inject
    public StandaloneAhcWSClient(AsyncHttpClient asyncHttpClient, Materializer materializer) {
        this.asyncHttpClient = asyncHttpClient;
        this.materializer = materializer;
    }

    @Override
    public Object getUnderlying() {
        return asyncHttpClient;
    }

    @Override
    public StandaloneAhcWSRequest url(String url) {
        return new StandaloneAhcWSRequest(this, url, materializer);
    }

    @Override
    public void close() throws IOException {
        asyncHttpClient.close();
    }

    CompletionStage<StandaloneWSResponse> execute(Request request) {
        final Promise<StandaloneWSResponse> scalaPromise = scala.concurrent.Promise$.MODULE$.apply();

        AsyncCompletionHandler<Response> handler = new ResponseAsyncCompletionHandler(scalaPromise);

        try {
            asyncHttpClient.executeRequest(request, handler);
        } catch (RuntimeException exception) {
            scalaPromise.failure(exception);
        }
        Future<StandaloneWSResponse> future = scalaPromise.future();
        return FutureConverters.toJava(future);
    }

    CompletionStage<StandaloneWSResponse> executeStream(Request request, ExecutionContext ec) {
        final Promise<StandaloneWSResponse> streamStarted = scala.concurrent.Promise$.MODULE$.apply();
        final Promise<Done> streamCompletion = scala.concurrent.Promise$.MODULE$.apply();

        Function<StreamedState, StandaloneWSResponse> f = state -> {
            Publisher<HttpResponseBodyPart> publisher = state.publisher();
            Publisher<HttpResponseBodyPart> wrap = new Publisher<HttpResponseBodyPart>() {
                @Override
                public void subscribe(Subscriber<? super HttpResponseBodyPart> s) {
                    publisher.subscribe(
                        new Subscriber<HttpResponseBodyPart>() {
                            @Override
                            public void onSubscribe(Subscription sub) {
                                s.onSubscribe(sub);
                            }

                            @Override
                            public void onNext(HttpResponseBodyPart httpResponseBodyPart) {
                                s.onNext(httpResponseBodyPart);
                            }

                            @Override
                            public void onError(Throwable t) {
                                s.onError(t);
                            }

                            @Override
                            public void onComplete() {
                                FutureConverters$.MODULE$.toJava(streamCompletion.future())
                                    .handle((d, t) -> {
                                        if (d != null) s.onComplete();
                                        else s.onError(t);
                                        return null;
                                    });
                            }
                        }
                    );
                }
            };

            return new StreamedResponse(this,
                state.statusCode(),
                state.statusText(),
                state.uriOption().get(),
                state.responseHeaders(),
                wrap,
                asyncHttpClient.getConfig().isUseLaxCookieEncoder());
        };

        asyncHttpClient.executeRequest(request, new DefaultStreamedAsyncHandler<>(f,
            streamStarted,
            streamCompletion
        ));
        return FutureConverters.toJava(streamStarted.future());
    }

    /**
     * A convenience method for creating a StandaloneAhcWSClient from configuration.
     *
     * @param ahcWSClientConfig the configuration object
     * @param materializer      an akka materializer
     * @return a fully configured StandaloneAhcWSClient instance.
     * @see #create(AhcWSClientConfig, AhcHttpCache, Materializer)
     */
    public static StandaloneAhcWSClient create(AhcWSClientConfig ahcWSClientConfig, Materializer materializer) {
        return create(
            ahcWSClientConfig,
            null /* no cache*/,
            materializer
        );
    }

    /**
     * A convenience method for creating a StandaloneAhcWSClient from configuration.
     *
     * @param ahcWSClientConfig the configuration object
     * @param cache             if not null, will be used for HTTP response caching.
     * @param materializer      an akka materializer
     * @return a fully configured StandaloneAhcWSClient instance.
     */
    public static StandaloneAhcWSClient create(AhcWSClientConfig ahcWSClientConfig, AhcHttpCache cache, Materializer materializer) {
        AhcLoggerFactory loggerFactory = new AhcLoggerFactory(LoggerFactory.getILoggerFactory());

        // Set up debugging configuration
        if (ahcWSClientConfig.wsClientConfig().ssl().debug().enabled()) {
            new DebugConfiguration(loggerFactory).configure(ahcWSClientConfig.wsClientConfig().ssl().debug());
        }

        // Configure the AsyncHttpClientConfig.Builder from the application.conf file...
        final AhcConfigBuilder builder = new AhcConfigBuilder(ahcWSClientConfig);
        final DefaultAsyncHttpClientConfig.Builder ahcBuilder = builder.configure();

        // Set up SSL configuration settings that are global..
        new SystemConfiguration(loggerFactory).configure(ahcWSClientConfig.wsClientConfig().ssl());

        // Create the AHC asyncHttpClient
        DefaultAsyncHttpClientConfig asyncHttpClientConfig = ahcBuilder.build();
        DefaultAsyncHttpClient defaultAsyncHttpClient = new DefaultAsyncHttpClient(asyncHttpClientConfig);

        AsyncHttpClient ahcClient;
        if (cache != null) {
            ahcClient = new CachingAsyncHttpClient(defaultAsyncHttpClient, cache);
        } else {
            ahcClient = defaultAsyncHttpClient;
        }
        return new StandaloneAhcWSClient(ahcClient, materializer);
    }

    ByteString blockingToByteString(Source<ByteString, ?> bodyAsSource) {
        try {
            return bodyAsSource
                .runFold(ByteString.createBuilder(), ByteStringBuilder::append, materializer)
                .thenApply(ByteStringBuilder::result)
                .toCompletableFuture()
                .get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    private static class ResponseAsyncCompletionHandler extends AsyncCompletionHandler<Response> {
        private final Promise<StandaloneWSResponse> scalaPromise;

        public ResponseAsyncCompletionHandler(Promise<StandaloneWSResponse> scalaPromise) {
            this.scalaPromise = scalaPromise;
        }

        @Override
        public Response onCompleted(Response response) {
            StandaloneAhcWSResponse r = new StandaloneAhcWSResponse(response);
            scalaPromise.success(r);
            return response;
        }

        @Override
        public void onThrowable(Throwable t) {
            scalaPromise.failure(t);
        }
    }
}
